Backup

noriyotcp opened the pull request:
Close #13 

ユーザーがメールアドレスを送信、そのアドレス宛にzipのダウンロードリンクを送る

TODOs
- [x] バックアップ用jsonがuser.jsonだけなので、他を作成する
- [x] zipファイルの置き場所きちんとする
- [x] ３日以上たったzipの削除
- [x] テスト書く
- [x] Backup#run -> #create
- [x] flashの文言
- [x] Download backup button
- [x] https://github.com/takeyuwebinc/gitfab2/pull/64/files/00d1cedcecf51e428aff1b5a8f7abb5c5dcf4df1#diff-993c64af0adada4c9e8097cbadb04521

noriyotcp committed:
Backup

Add task to delete zip files

引数はdirpathとday
dirpath以下のファイルが対象
dayに3を与えると本日から3日以上経っているファイルを削除する

Add spec against backup.rake

itkrt2y commented on app/models/user/backup.rb:
```diff
@@ -0,0 +1,97 @@
+
+# frozen_string_literal: true
+
+class User::Backup < User
```
「バックアップのユーザー」って概念はおかしいので、Backup Serviceにすべきですね。
参考にしたであろう `User::PasswordAuth` は概念的にちょっと微妙なところかもしれませんが、でも、「パスワード認証をしているユーザー」って概念はありうるのであり。

```rb
# app/services/backup.rb
class Backup
  attr_reader :user

  def initialize(user)
    @user = user
  end

  def run
    # メインの処理
  end
end
```

itkrt2y commented on app/models/user/backup.rb:
```diff
@@ -0,0 +1,97 @@
+  end
+
+  def path_to_zip
+    "#{Rails.root}/tmp/#{name}_backup.zip"
```
今回は関係ないでしょうけど、windows環境の人に優しくするなら、パスを作るときは `/` でつなげていくのではなくて `Rails.root.join("tmp", "{name}_backup.zip")` としたほうがいいです。

itkrt2y commented on app/models/user/backup_service.rb:
```diff
@@ -0,0 +1,98 @@
+
+# frozen_string_literal: true
+
+class User::BackupService < User
```
https://github.com/takeyuwebinc/gitfab2/pull/64#discussion_r225464521

noriyotcp replied:
すごい勘違いしてました。たった今force pushして戻しました

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+
+  def path_to_zip
+    Rails.root.join("#{backup_zip_dir}", "#{@user.name}_backup.zip")
```
```suggestion
    backup_zip_dir.join("#{@user.name}_backup.zip")
```

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+
+  def dir_to_zip
+    Rails.root.join('tmp', 'backup', "#{@user.name}")
```
```suggestion
    Rails.root.join('tmp', 'backup', @user.name)
```

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,79 @@
+    end
+
+    def generate_zip_file
+      FileUtils.mkdir("#{backup_zip_dir}")
```
```suggestion
      FileUtils.mkdir(backup_zip_dir)
```

noriyotcp committed:
Run backup.rake via whenever

Add whenever settings for Capistrano

itkrt2y commented on app/controllers/users_controller.rb:
```diff
@@ -84,6 +84,22 @@ def update_password
+    return if params[:download_backup_token].blank?
+
+    user = User.active.find_by(name: params[:user_id])
+    send_data(File.read(Backup.new(user).path_to_zip), filename: "#{user.name}_backup.zip")
```
```suggestion
    send_data(Backup.new(user).path_to_zip.read, filename: "#{user.name}_backup.zip")
```

itkrt2y replied:
`"#{user.name}_backup.zip"` は `Backup` クラスの中にも同じ処理があってDRYじゃないので、 `zip_filename` とか作るべき

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,79 @@
+    Rails.root.join(backup_zip_dir, "#{@user.name}_backup.zip")
+  end
+
+  def dir_to_zip
```
詳細に理解できてないですけれども、 `backup_zip_dir` と `dir_to_zip` は明らかに責務が重複してるので、まとめるなりもっと適切な名前をつけるなりした方がいいです。
また、 `backup_zip_dir`  というメソッド名はBackupクラスの中なのだから `backup_` なのは当たり前なので、このprefixは不要。

itkrt2y replied:
あと、 `dir_to_zip` よりも `zip_dir` って名付けの方が一般的。
もしくは `output_dir` か。

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+
+  def path_to_zip
+    Rails.root.join(backup_zip_dir, "#{@user.name}_backup.zip")
```
`backup_zip_dir` の返り値は `Pathname` オブジェクトなので、 そのままjoin可能

```suggestion
    backup_zip_dir.join("#{@user.name}_backup.zip")
```

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,79 @@
+  private
+
+    def generate_json_files
+      FileUtils.mkdir_p("#{dir_to_zip}/user")
```
ここでさらにpathをjoinしちゃうと、上記のパスを作ってるメソッドはなんのためにいるのか曖昧になります

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,79 @@
+          end
+      }
+
+      [user_hash, identities_hash, groups_hash].reduce(:merge)
```
`reduce(:merge)` とかするくらいなら、最初から全部Hashでくるんじゃえばいいような

```rb
def user_data
  {
    source: url,
    name: user.name,
    # 中略
    identities: user.identitties.map do |identity|
      {
        provider: identity.provider
        # 中略
      }
    end,
    groups: user.groups.map do |group|
      {
        name: group.name,
        # 中略
      }
    end
  }
end
```

itkrt2y replied:
ついでに、このクラスでは `@user` は絶対にいる想定で作られているので（作るべきなので）、引数に `user` を渡す必要はないです

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,79 @@
+
+class Backup
+  def initialize(user)
+    @user = user
```
外から `user` を見たいわけじゃないので必須ではないですが、 `attr_reader :user` にしておいて `@user` を直接呼ぶのは `initialize` 内での初期化のみとしておくと、`@user` がどこからも書き換えられていないことをより強く明示できます

itkrt2y replied:
あともうひとつ付け加えておくと、インスタンス変数って未定義のを読んでもエラーにならず `nil` が返ってくるんですけど、メソッドに包むとtypoでバグるのを防げる。
外から見えるのが嫌なら `private def user` と `private` にすればOK

itkrt2y replied:
以下でよかった
```
private
  attr_reader :user
```

itkrt2y commented on Gemfile:
```diff
@@ -27,6 +27,8 @@ gem 'slack-notifier'
 gem 'sprockets-commoner'
 gem 'truncate_html'
+gem 'rubyzip'
+gem 'whenever', require: false
```
Gemを変更する際は、 `gemfiles/rails-master.gemfile` にも修正お願いします

noriyotcp committed:
Introduce delayed_job

Use perform_later

noriyotcp committed:
Render 403 when the user is forbidden to backup

noriyotcp committed:
Add test against BackupJob

noriyotcp committed:
Add test against UserMailer#backup

itkrt2y commented on spec/jobs/backup_job_spec.rb:
```diff
@@ -0,0 +1,25 @@
+RSpec.describe BackupJob, type: :job do
+  ActiveJob::Base.queue_adapter = :test
```
これは `config/environments/test.rb` に書くべきでしょう

itkrt2y commented on app/views/users/edit.html.slim:
```diff
@@ -45,6 +45,17 @@
+    = form_with(url: user_backup_path(current_user), method: :patch, local: true) do |f|
+      .form-group
+        = f.label :email, "メールアドレス"
+        = f.email_field :email, class: "form-control"
```
メールアドレス、まだユーザー登録に必須じゃなかった時代は入力フォームを作る話をしてましたが、今は必須なので入力させる必要はないです

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,204 @@
+
+  private
+
+    def input_dir
```
`input_dir` じゃなくて、 `json_output_dir` と `zip_output_dir` の2つの出力先があるんですよね？

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,204 @@
+      zf.write
+    end
+
+    def generate_user_hash
```
`generate` っていうと何か副作用があるっぽく見えますけれども実際はデータを返してるだけなので、Rubyのお約束的には `user_data` とかで十分です

itkrt2y commented on app/models/zip_file_generator.rb:
```diff
@@ -0,0 +1,57 @@
+require 'zip'
```
公式ドキュメントからコピペするのはいいんですけど、そういう時はコピー元をコメントで残して欲しいです。
また、コメントに書く時はbranch名を使わず、コミットハッシュで参照する。

https://github.com/rubyzip/rubyzip/blob/d07b13a6cf0a413e010c48879aebd9576bfb5f68/README.md#zipping-a-directory-recursively

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,204 @@
+        url: @user.url,
+        location: @user.location,
+        email: @user.email,
+        avatar: "https://fabble.cc#{@user.avatar.url}",
```
`"https://fabble.cc"` ての、現在真っ当な用意方法がないのであれですが、せめて `base_url` に切り出してDRYにするとか、pathをつなげる時は `URI.join` を使うとかしないと危なすぎます

itkrt2y replied:
こう言う感じ

```rb
private

def base_url
  "https://fabble.cc"
end

def image_url(path)
  URI.join(base_url, path).to_s
end
```

itkrt2y replied:
こういう感じ

```rb
private

def base_url
  "https://fabble.cc"
end

def image_url(path)
  URI.join(base_url, path).to_s
end
```

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,204 @@
+        projects: @user.project_comments.map do |comment|
+          {
+            # TODO: project_commentへのアンカーを入れたのだが、要確認
+            source: Rails.application.routes.url_helpers.project_url(comment.project.owner, comment.project, anchor: "project-comment-#{comment.id}", host: "https://fabble.cc"),
```
classの冒頭で `include Rails.application.routes.url_helpers` ってしちゃうといいです

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,204 @@
+      {
+        projects: @user.project_comments.map do |comment|
+          {
+            # TODO: project_commentへのアンカーを入れたのだが、要確認
```
これでいいと思います

itkrt2y commented on config/deploy.rb:
```diff
@@ -6,6 +6,7 @@
 set :rvm_ruby_version, '2.5.1'
 set :assets_roles, [:web, :app]
 set :npm_flags, '--silent --no-progress' # Not use --production to install devDependencies
+set :whenever_command, ->{ [:bundle, :exec, :whenever] }
```
`set :whenever_command, "bundle exec whenever"` でいいと思うんですけど、lambda使ってるの何ででしょう？
https://github.com/javan/whenever#capistrano-integration

noriyotcp replied:
ここを参考にしてました

https://github.com/javan/whenever/blob/6e69dd8a6b3e2a8f4b2911b4efa5aab65cdc9dcb/lib/whenever/capistrano/v3/tasks/whenever.rake#L46

itkrt2y replied:
どうせ内部的にStringに変換して二度手間ですし、最初からStringで
```suggestion
set :whenever_command, "bundle exec whenever"
```

itkrt2y commented on config/routes.rb:
```diff
@@ -45,6 +45,8 @@
   resources :users, except: [:show, :edit, :update, :destroy] do
     resources :memberships, only: [:index, :update, :destroy]
     patch :update_password
+    patch :backup
```
`post` だと思います

itkrt2y commented on config/schedule.rb:
```diff
@@ -0,0 +1,4 @@
+every 1.day, at: ['5:00 am'] do
```
自信ないですが、whenever越しでもcronはRailsアプリのタイムゾーンを考慮しないんじゃないかな......？

noriyotcp replied:
考慮しないみたいですね 😩 
https://github.com/javan/whenever/issues/662

サーバー側のタイムゾーンを変えないのであれば、utcにパースして設定するのが良いか

https://github.com/javan/whenever/issues/662#issuecomment-419024516

itkrt2y replied:
UTCにparseする処理を追加するくらいなら、最初からUTCで書いてコメントにJSTの時刻を書くのがいいと思います

itkrt2y commented on app/controllers/users_controller.rb:
```diff
@@ -84,6 +84,24 @@ def update_password
+    user = User.active.find_by(name: params[:user_id])
+    if can?(:backup, user)
+      BackupJob.perform_later(user, params[:email])
+      redirect_to edit_user_path, flash: { success: "バックアップ用のメールを送信しました" }
```
「メールを送信しました」はちょっと違いますね。即送ってるわけじゃないですし。

ちなみにTwitterはこんな感じ
> 全ツイート履歴を含むファイルをリクエストできます。ダウンロードの準備ができたらメールでリンクをお送りします。

「バックアップを開始しました。バックアップファイルのダウンロードの準備ができたらメールでリンクをお送りします。」って感じでいいんじゃないでしょうか

itkrt2y commented on lib/tasks/backup.rake:
```diff
@@ -0,0 +1,13 @@
+  desc 'Delete backup files under the directory'
+  task :delete, [:dirpath, :day] => :environment do |_task, args|
+    args.with_defaults(day: 3)
+    zip_files = Dir.glob("#{args[:dirpath]}/*")
```
バックアップ出力先をコロコロ変えるようにするわけじゃないのにパスを渡す形式にする必要ないです

itkrt2y replied:
バックアップ出力先は基本変えないので、パスを渡す形式にする必要ないです。むしろパスを渡す形式になってると、cliから叩くのがすごく面倒になるのでマイナス

noriyotcp committed:
Refactoring

Change patch to post

Remove email_field and do not pass email from form
- email登録必須となったため、フィールドへの入力は不要になった
- user.emailを使用する

Pass user only to BackupJob.perform
Do not pass email to UserMailer#backup

Add a reference link to zip_file_generator.rb

itkrt2y commented on lib/tasks/backup.rake:
```diff
@@ -0,0 +1,13 @@
+namespace :backup do
+  desc 'Delete backup files under the directory'
+  task :delete, [:dirpath, :day] => :environment do |_task, args|
+    args.with_defaults(day: 3)
```
日付を渡せる必要ないんじゃないですかね。使わないだろうし、使うなら使うで `rails console` で以下のスクリプトを書いてもいいし

noriyotcp committed:
Refactor services/backup.rb

Include Rails.application.routes.url_helpers
Rename input_dir to json_output_dir, output_dir to zip_output_dir
Remove generate_ from methods those returns hash
Add base_url which returns 'https://fabble.cc'
Add image_url which generates base_url + image url via URI.join
Add private attr_reader :user

itkrt2y commented on lib/tasks/backup.rake:
```diff
@@ -0,0 +1,13 @@
+      if s.mtime.to_date <= args[:day].to_i.days.ago
+        File.delete(zip)
+      end
+    end
```
削除処理は `Backup` クラスが持ってた方がいい気がします。
`Backup.delete_old_files` とか作って、現在の `Backup#run` は `Backup#create` とかにリネームする

itkrt2y commented on app/controllers/users_controller.rb:
```diff
@@ -84,6 +84,24 @@ def update_password
+  end
+
+  def download_backup
+    return if params[:download_backup_token].blank?
```
これでは何でもいいから適当なtokenをURLにつければダウンロードできてしまいますよね。
tokenをどこかに保存しておいて、URLにつけられたtokenと保存されているもののすり合わせをしないと。

itkrt2y replied:
token形式にこだわる必要もなくて、"Backup"ボタンの下にバックアップ済みの場合にのみ"Download Backup"ボタンを表示するようにして、バックアップファイルができたらそこへのリンク（ `https://fabble.cc/user/edit#backup` とかになると思う）を貼ったメールを送るのでもいい。
その場合は、"Backup"ボタンを"Create Backup"って文言に変えるのがいいでしょう。

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,204 @@
+    def generate_json_files
+      # input_dirをルートとして階層を作る
+      # ZipFileGeneratorに渡して、再帰的にzipしてもらう
+      unless Dir.exist?(input_dir)
```
全体的にですが、 `input_dir` や `output_dir` って奴は `Pathname` オブジェクトであることを意識するともっと綺麗に書けます。
例えばここは `input_dir.exist?` でいい。

noriyotcp committed:
Add Backup.delete_old_files and call it from backup:delete rake task

rake taskに引数を渡すのはやめる

noriyotcp committed:
Rename Backup#run to Backup#create

noriyotcp committed:
Refactor Backup#generate_json_files and #generate_zip_file using Pathname class

noriyotcp committed:
Download zip file by clicking 'Download Backup' button

backup zipが作成されたら、メールを送る
backup zipが存在するときのみ、Download Backupボタンの表示
メールのリンクは、そのDownload Backupへのアンカー付き

Change flash message on users#backup

Refactor backup_job.rb and user_mailer.rb
tokenを生成するのをやめた

Delete spec/lib/tasks/backup_spec.rb

itkrt2y commented on spec/lib/tasks/backup_spec.rb:
```diff
@@ -0,0 +1,15 @@
+
+  it do
+    expect(Backup).to receive(:delete_old_files).once
+    @rake['backup:delete'].invoke
```
さすがにこのテストは必要ないです

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,226 @@
+  end
+
+  def self.delete_old_files
+    zip_files = Dir.glob("#{Rails.root.join('tmp', 'backup', 'zip', '*')}")
```
```suggestion
    zip_files = Dir.glob(Rails.root.join('tmp', 'backup', 'zip', '*'))
```

itkrt2y committed:
Omit String interpolation

Co-Authored-By: noriyotcp <noriyo.akita@gmail.com>

itkrt2y commented on app/models/zip_file_generator.rb:
```diff
@@ -0,0 +1,58 @@
+
+  # Zip the input directory.
+  def write
+    entries = Dir.entries(@input_dir) - %w(. ..)
```
```suggestion
    entries = Dir.children(@input_dir)
```

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,230 @@
+  end
+
+  def self.zip_exist?(user)
+    Rails.root.join('tmp', 'backup', 'zip', "#{user.name}_backup.zip").exist?
```
```suggestion
    new(user).path_to_zip.exist?
```

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,230 @@
+    "#{user.name}_backup.zip"
+  end
+
+  def path_to_zip
```
他のメソッド名とかとの一貫性や単純な読みやすさから言うと、 `zip_path` の方が好ましいです

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,230 @@
+    attr_reader :user
+
+    def json_output_dir
+      Rails.root.join('tmp', 'backup', user.name)
```
`Rails.root.join('tmp', 'backup')` まででメソッド作って使いまわしたいです

itkrt2y committed:
Change lambda to String

Co-Authored-By: noriyotcp <noriyo.akita@gmail.com>

itkrt2y committed:
Use Dir.children instead of Dir.entries

Co-Authored-By: noriyotcp <noriyo.akita@gmail.com>

itkrt2y committed:
Use Backup#path_to_zip instead of Pathname

Co-Authored-By: noriyotcp <noriyo.akita@gmail.com>

noriyotcp committed:
Rename Backup#path_to_zip to #zip_path

for consistency and readability

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,230 @@
+  end
+
+  def self.zip_exist?(user)
+    new(user).path_to_zip.exist?
```
自分で提案したのをさらに指摘しちゃってすみませんけど、以下の方が回りくどくなくていい気がしました

```rb
def zip_exist?
  path_to_zip.exist?
end

Backup.new(user).zip_exist?
```

noriyotcp committed:
Refactor services/backup.rb

Add backup_dir which indicates 'tmp/backup'
Change zip_exist? from class method to instance method

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+
+  def self.delete_old_files
+    zip_files = Dir.glob(Rails.root.join('tmp', 'backup', 'zip', '*'))
+    zip_files.each do |zip|
```
`zip_output_dir.children.each do |zip|` でいけそう

itkrt2y replied:
`each_child` ってのがあった
https://ruby-doc.org/stdlib-2.1.3/libdoc/pathname/rdoc/Pathname.html#method-i-each_child

```rb
zip_output_dir.each_child do |zip|
```

noriyotcp replied:
``undefined local variable or method `zip_output_dir' for Backup:Class``

`zip_output_dir`, `backup_dir` はユーザーに関係がないので、`private_constant` にして内部で使い回すのはどうかなと

noriyotcp replied:
やってみました
https://github.com/takeyuwebinc/gitfab2/pull/64/commits/331dce012ab06ce90268ce2b4e5e589f92944c51

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+  def create
+    generate_json_files
+    generate_zip_file
+    FileUtils.rm_rf(json_output_dir) # cleanup json directories
```
この行も `remove_json_files` みたいな名前のメソッドに包むと、このメソッドでやっていることがより分かりやすくなります

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+      FileUtils.copy(user.avatar.file.file, json_output_dir.join('user', 'avatar'))
+      user.projects.each do |project|
+        copy_project_contents(project)
+      end
```
"Generate JSON files"に画像などのコピーが入っていると責務が過剰ですね。
画像などの準備もメソッドに切り出した上で `#create` から呼ばれるべき

```rb
def create
  # 各種メソッド名は適当
  generate_json_files
  prepare_media
  generate_zip_file
  cleanup_tmp
end
```

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+        recipes: user.card_comments.map do |comment|
+          {
+            # TODO: source: https://fabble.cc/card_comments/323
+            # 上記のようになっているが要確認
```
このコメントはレビュアーも確認してくれってことですか？ それとも自分へのメモ？

noriyotcp replied:
すみません、両方です 🙇 
`card_comment` 個別のurlにはアクセスできないので、どうしたものかなと
`project_comment` のようにアンカーをつければ良いのか 🤔

itkrt2y replied:
アンカーでいいと思います

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+    zip_files = Dir.glob(Rails.root.join('tmp', 'backup', 'zip', '*'))
+    zip_files.each do |zip|
+      s = File::Stat.new(zip)
+      if s.mtime.to_date <= 3.days.ago
```
対応は必須ではないですが、rake taskの時はわざわざ時間指定できる必要なかったですけど、メソッドにするなら引数渡しをした方がテスタブルだしよかったりもします。

```
def self.delete_old_files(before = 3.days.ago)
  if s.mtime.to_date <= before
end

itkrt2y replied:
対応は必須ではないですが、バッチ処理の定義としてはわざわざ時間指定を出来る必要はなかったですけど、メソッドにするなら引数渡しをした方がテスタブルだしよかったりもします。

```
def self.delete_old_files(before = 3.days.ago)
  if s.mtime.to_date <= before
end

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+              created_at: card.created_at.iso8601,
+              media: card.figures.map { |figure| image_url(figure.content.url) },
+              youtube: card.figures.find { |figure| figure.link.present? }&.link,
+              material: card.attachments.select { |a| a.kind == 'material' },
```
ここ以下の `card.attachments` の並び、オブジェクトを指定してるだけですからダメですよね。
`card.attachments.where(kind: 'material').map{ |a| image_url(a.content_url) }` みたいな感じになるのでは

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+                  created_at: annotation.created_at.iso8601,
+                  media: annotation.figures.map { |figure| image_url(figure.content.url) },
+                  youtube: annotation.figures.find { |figure| figure.link.present? }&.link,
+                  material: annotation.attachments.select { |a| a.kind == 'material' },
```
こっちの `attachments` も

itkrt2y commented on app/views/user_mailer/backup.html.erb:
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @user.email %>!</p>
```
ここでは `email` よりも `name` の方が適切でしょう

itkrt2y commented on app/models/zip_file_generator.rb:
```diff
@@ -0,0 +1,58 @@
+
+    def recursively_deflate_directory(disk_file_path, zipfile, zipfile_path)
+      zipfile.mkdir zipfile_path
+      subdir = Dir.entries(disk_file_path) - %w(. ..)
```
```suggestion
      subdir = Dir.children(disk_file_path)
```

itkrt2y commented on spec/services/backup_spec.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  after(:each) do
+    FileUtils.rm(zip_path, force: true)
```
https://github.com/takeyuwebinc/gitfab2/pull/64/files#r228731729 を実装すると、ついでにここで使える

noriyotcp committed:
Add remove_json_files

noriyotcp committed:
Extract copying user contents to copy_contents method

noriyotcp committed:
Fix for attachments

attachmentのcontent.urlはnilの場合もあるのでimage_urlでguard clauseを入れている

noriyotcp committed:
Use user name instead of email

itkrt2y committed:
Use Dir.children instead of Dir.entries

Co-Authored-By: noriyotcp <noriyo.akita@gmail.com>

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,240 @@
+    end
+
+    def copy_contents
+      json_output_dir.join('user', 'avatar').mkpath
```
`avatar_dir = json_output_dir.join('user', 'avatar')` と定義して使い回す

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,240 @@
+              created_at: card.created_at.iso8601,
+              media: card.figures.map { |figure| image_url(figure.content.url) },
+              youtube: card.figures.find { |figure| figure.link.present? }&.link,
+              material: card.attachments.where(kind: 'material').map { |a| image_url(a.content_url).to_s },
```
`image_url` の中で `to_s` してるからここではいらないですよね

noriyotcp committed:
Add anchor to card_comment

noriyotcp committed:
Pass N.days.ago to Backup.delete_old_files

noriyotcp committed:
Change zip_output_dir and backup_dir to private constants

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+
+  BACKUP_DIR = Rails.root.join('tmp', 'backup')
+  ZIP_OUTPUT_DIR = BACKUP_DIR.join('zip')
+  private_constant :BACKUP_DIR, :ZIP_OUTPUT_DIR
```
なんでprivate？

noriyotcp replied:
外からアクセスする必要がなく、内部で使い回したいだけなのでプライベートにしました

itkrt2y replied:
外からアクセスすることはないですけど、外部に公開されて困るインターフェースじゃないですよね？

itkrt2y replied:
まあ、いいっちゃいいんですけども

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,234 @@
+    generate_json_files
+    copy_contents
+    generate_zip_file
+    remove_json_files # cleanup json directories
```
抽象化するためにメソッドに切り出してそれらしい名前をつけたのですから、コメントは不要
```suggestion
    remove_json_files
```

itkrt2y commented:
昨日時点のattachmentなどの実装内容からして全機能の確認をしていないと思うので、全機能の動作確認が出来たらもう一度レビュー待ちタグをつけてください。
そうしたら僕も最終確認をします

noriyotcp commented:
https://github.com/takeyuwebinc/gitfab2/blob/97b26784786fa86451cd514f8ff18d07bc40e23c/app/services/backup.rb#L143-L146

動作確認中気づいたのですが、`attachment`は`content`が無い場合もありますね？

itkrt2y commented:
> 動作確認中気づいたのですが、`attachment`は`content`が無い場合もありますね？

`compact` すれば

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,236 @@
+            created_at: identity.created_at.iso8601
+          }
+        end,
+        groups: user.groups.map do |group|
```
```suggestion
        groups: user.groups.active.map do |group|
```

itkrt2y commented on app/services/backup.rb:
```diff
@@ -0,0 +1,236 @@
+    end
+
+    def projects_hash
+      user.projects.map do |project|
```
```suggestion
      user.projects.active.map do |project|
```

itkrt2y committed:
Wrap groups/projects with methods

itkrt2y committed:
Show backup's created_at
